### PR TITLE
Make GraphTile own their memory

### PR DESCRIPTION
**Pull request by kkaefer**
`GraphTile`s currently own their memory if they're loaded from disk as plain files, or if they're compressed. However, uncompressed tiles in a tar file must hope that the underlying mapped memory won't get unmapped.

This PR moves towards a model where all `GraphTile` objects explicitly own their memory. Simultaneously, `GraphTile`s can't be copied anymore. ~The various *Cache objects now use `shared_ptr`s for ownership as well.~

## Tasklist

 - [ ] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Generally use squash merge to rebase and clean comments before merging
 - [ ] Update the [changelog](CHANGELOG.md)

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

**Comment by kevinkreiser** [comment]
I still cant remember why we didnt do it this way in the first place. Thank you for this huge cleanup!

File: valhalla/baldr/graphmemory.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#ifndef VALHALLA_BALDR_GRAPHMEMORY_H_
+#define VALHALLA_BALDR_GRAPHMEMORY_H_
+
+#include <memory>
```
Comment (mandeepsandhu) [inline valhalla/baldr/graphmemory.h:4]: Whats this include for?

Comment (kkaefer) [reply valhalla/baldr/graphmemory.h:4]: You're right, it's not used and should be removed.

File: test/CMakeLists.txt [hunk 2]
```diff
@@ -69,10 +69,10 @@ set_target_properties(build_timezones PROPERTIES FOLDER "Tests")
 
 add_custom_command(OUTPUT ${CMAKE_BINARY_DIR}/test/data/utrecht_tiles/0/003/196.gph
   COMMAND ${CMAKE_COMMAND} -E make_directory test/data/utrecht_tiles/
-  COMMAND ${CMAKE_BINARY_DIR}/valhalla_build_tiles
+  COMMAND $<TARGET_FILE:valhalla_build_tiles>
```
Comment (kevinkreiser) [inline test/CMakeLists.txt:72]: @kkaefer seems like windows build system doesnt like this change. 

```
CMake Error at test/CMakeLists.txt:70 (add_custom_command):
  Error evaluating generator expression:

    $<TARGET_FILE:valhalla_build_tiles>

  No target "valhalla_build_tiles"
```

**Comment by kevinkreiser** [comment]
@kkaefer i had a go at cleaning this up for CI. with any luck all builds will pass

File: test/CMakeLists.txt [hunk 3]
```diff
@@ -69,10 +69,10 @@ set_target_properties(build_timezones PROPERTIES FOLDER "Tests")
 
 add_custom_command(OUTPUT ${CMAKE_BINARY_DIR}/test/data/utrecht_tiles/0/003/196.gph
   COMMAND ${CMAKE_COMMAND} -E make_directory test/data/utrecht_tiles/
-  COMMAND $<TARGET_FILE:valhalla_build_tiles>
+  COMMAND ${CMAKE_BINARY_DIR}/valhalla_build_tiles
```
Comment (kkaefer) [inline test/CMakeLists.txt]: I made this change because e.g. Xcode uses per-buildtype output directories and I couldn't build various targets because the path above is incorrect. However, I agree that this change is unrelated to GraphTiles, so we can move it to a different patch. It seems like the `valhalla_build_tiles` target doesn't exist on Windows at all.

**Comment by kkaefer** [comment]
Thanks for bring it up to date with master! Do you see any need for further changes?

**Comment by kevinkreiser** [comment]
@kkaefer I dont think so we should just benchmark it to be sure there is no measureable impact.

**Comment by dnesbitt61** [comment]
for what is worth I ran a benchmark using a bicycle route from Frederick, MD to Pittsburgh, PA
mmap:  master 822 ms  branch  844 ms
compressed tiles: master 820 ms  branch 850 ms

**Comment by kkaefer** [comment]
@dnesbitt61 could you please share how you're conducting the benchmark? Are these numbers representative of other workloads as well? +@kevinkreiser, is there a more formal way of running benchmarks?

One optimization opportunity is to replace the `std::unique_ptr<const GraphMemory> memory_;` member of GraphTile with `std::aligned_storage` based type-erasing storage that is big enough for a `std:shared_ptr`. This would reduce one dereference. However, we're not typically accessing the `memory_` member; it mostly serves an ownership function, so I don't think this would bring massive improvements.

I'm not sure why this branch is slower; it removes several copies. Otoh, std::shared_ptr uses locking in the control block for thread safety, which could contribute to the runtime increase.

**Comment by dnesbitt61** [comment]
The benchmark I run is using _valhalla_run_route_ application with the _multi-run_ option. This isolates the pathfinding method (bidirectional A* GetBestPath) and runs it sequentially. I ran it using 100 iterations. Timing for the first pass is ignored and it outputs the average time of the successive runs. This removes the time spent loading tiles into cache (esp. important for compressed tiles) and focuses on just the bidirectional A* algorithm. Running a bicycle route basically touches most edges in a tile, but limits how many tiles are actually needed. I used to run a benchmark driving (auto) route from Lancaster, PA to San Diego, CA and used this as a rough measure of how performance of bidirectional A* changed over time. I couldn't run an auto route since my data is optimized for bicycle (or pedestrian) routing by not adding hierarchies and shortcuts.

The command is here (fill in your own config file):
./valhalla_run_route --config ../../conf/planet01_20.json --multi-run 100 -j '{"locations":[{"lat":39.310012,"lon":-77.463554,"type":"break"},{"lat":40.442265,"lon":-80.007329,"type":"break"}],"costing":"bicycle","costing_options":{"bicycle":{"bicycle_type":"Road","use_roads":"0.5","cycling_speed":"25.0","use_hills":"0.5"}}}'

**Comment by kkaefer** [comment]
Thanks, I replicated your benchmark with the removal of shared_ptrs in the latest commit and got these numbers for tarballs:

- this branch: 340, 351, 340, 341
- master branch: 348, 328, 339, 332

**Comment by kkaefer** [comment]
@kevinkreiser I still think this change would be useful to have in Valhalla. Is there anything that is blocking a merge, apart from resolving conflicts?

**Comment by kevinkreiser** [comment]
@kkaefer i agree. i think if we resolve the conflicts i can do some performance testing and we can get it shipped. sorry for the delay!

**Comment by averkhaturau** [comment]
@kevinkreiser, @kkaefer, this branch is rebased an tests fixed. Please review and merge.

**Comment by kevinkreiser** [comment]
Good news. As was the case before, I can measure no discernable performance difference between this branch and master. So we just need to do a code review before its shippable. Thanks for doing the long haul @kkaefer and @averkhaturau

**Comment by averkhaturau** [comment]
Thank you, @kevinkreiser . Please feel free to merge it if no code changes required.

File: test/graphreader.cc [hunk 4]
```diff
@@ -146,33 +156,33 @@ TEST(SimpleCache, Clear) {
...
   // Check if inserted values are correct
 
-  const auto* returned2 = cache.Get({300, 1, 0});
+  const auto returned2 = cache.Get({300, 1, 0});
   CheckGraphTile(returned2, id2, 200);
-  const auto* returned1 = cache.Get({100, 2, 0});
+  const auto returned1 = cache.Get({100, 2, 0});
   CheckGraphTile(returned1, id1, 123);
-  const auto* returned3 = cache.Get({1000, 0, 0});
+  const auto returned3 = cache.Get({1000, 0, 0});
```
Comment (kevinkreiser) [inline test/graphreader.cc:185]: these are pointers, i know type deduction will include the pointer in `auto` but its less illustrative of what is going on to remove the pointer. can we revert these please

Comment (kevinkreiser) [reply test/graphreader.cc:185]: seems this whole file has similar changes it would be nice to revert all of them

File: test/graphtilebuilder.cc [hunk 5]
```diff
@@ -30,7 +30,8 @@ class test_graph_tile_builder : public GraphTileBuilder {
 void assert_tile_equalish(const GraphTile& a,
```
Comment (kevinkreiser) [inline test/graphtilebuilder.cc:30]: maybe im misreading it but it looks like non of these changes were needed in this test

File: test/graphreader.cc (hunk 4 above)
Comment (averkhaturau) [reply test/graphreader.cc:185]: Get returns boost::intrusive_ptr<const baldr::GraphTile>, so this is not a raw pointer.

File: test/test.cc [hunk 6]
```diff
@@ -21,8 +22,39 @@
 
 #include "microtar.h"
 
+namespace valhalla {
```
Comment (kevinkreiser) [inline test/test.cc:25]: did we really need to add this additional namespace. im going to guess that it causes a lot of the diffs in the tests. was the point to remove a potential namespace collision/merging with other libraries since its a very common name (`test`)?

File: test/traffictile.cc [hunk 7]
```diff
@@ -2,6 +2,16 @@
 
 #include "baldr/traffictile.h"
 
+namespace {
+class IrresponsibleGraphMemory : public valhalla::baldr::GraphMemory {
```
Comment (kevinkreiser) [inline test/traffictile.cc:6]: ouch!? maybe `unmanaged` would be better convey the sentiment here?

File: test/test.cc (hunk 6 above)
Comment (kevinkreiser) [reply test/test.cc:25]: or maybe it doesnt make lots of diffs because most tests say `using namespace valhalla;`?

**Comment by kevinkreiser** [comment]
@averkhaturau a couple nitpicks but otherwise looks good! ill merge if you can do a couple cleanups

File: test/graphreader.cc (hunk 4 above)
Comment (kevinkreiser) [reply test/graphreader.cc:185]: not in this pr it doesn :smile:

File: test/test.cc (hunk 6 above)
Comment (averkhaturau) [reply test/test.cc:25]: it was an attempt to minimize changes, but better to remove this namespace.

File: test/graphreader.cc (hunk 4 above)
Comment (averkhaturau) [reply test/graphreader.cc:185]: Ok, will add *

File: test/graphtilebuilder.cc (hunk 5 above)
Comment (averkhaturau) [reply test/graphtilebuilder.cc:30]: ok

**Review by kevinkreiser — DISMISSED** [review]

**Comment by kevinkreiser** [comment]
I'm going to add a changelog entry and merge it! Thanks again for the long haul!

**Review by kevinkreiser — APPROVED** [review]
